src/MSBuild/XMake.cs

```diff
@@ -3456,13 +3456,27 @@ private static bool DoesEnvironmentSupportLiveLogger()
 
         private static void ProcessLiveLogger(
             bool noConsoleLogger,
+            List<DistributedLoggerRecord> distributedLoggerRecords,
+            int cpuCount,
             List<ILogger> loggers)
         {
-            // Check for flags and env variables
             if (!noConsoleLogger)
             {
-                LiveLogger l = new LiveLogger();
-                loggers.Add(l);
+                // A central logger will be created for both single proc and multiproc.
+                LiveLogger logger = new LiveLogger();
+
+                // Check to see if there is a possibility we will be logging from an out-of-proc node.
+                // If so (we're multi-proc or the in-proc node is disabled), we register a distributed logger.
+                if (cpuCount == 1 && !Traits.Instance.InProcNodeDisabled)
+                {
+                    loggers.Add(logger);
+                }
+                else
+                {
+                    // For performance, register this logger using the forwarding logger mechanism.
+                    DistributedLoggerRecord forwardingLoggerRecord = CreateForwardingLoggerRecord(logger, string.Join(";", LiveLogger.ConfigurableForwardingLoggerParameters), LoggerVerbosity.Quiet);
```

Comment:
Ew, this is much harder than I wish it was.

Reply:
You thought that we might create a custom filtering implementation? I thought about that, but I noticed that current filtering needs are satisfied with standard ConfigurableForwardingLogger so I reused it. As for logic in this function, it copies logic for console logger in sense of when to make a central logger and when to make it distributed.

Reply:
Yeah, I think both decisions (use ConfigurableForwardingLogger, register in the same way as the `-dl` switch) are good. But It's uglier than I hoped.

Reply:
The intended way to use ConfigurableForwardingLogger from our own code seems quite ugly to me, yeah.